modules/dnn/src/layers/elementwise_layers.cpp

```diff
@@ -813,20 +813,173 @@ struct BaseDefaultFunctor : public BaseFunctor
+        _vexp_mm = v_shl(_vexp_mm, 23);
+
+        _vexp_x = v_fma(_vexp_, _vexp_C1_f32, _vexp_x);
+        _vexp_x = v_fma(_vexp_, _vexp_C2_f32, _vexp_x);
+        _vexp_xx = v_mul(_vexp_x, _vexp_x);
+
+        _vexp_y = v_fma(_vexp_x, _vexp_p0_f32, _vexp_p1_f32);
+        _vexp_y = v_fma(_vexp_y, _vexp_x, _vexp_p2_f32);
+        _vexp_y = v_fma(_vexp_y, _vexp_x, _vexp_p3_f32);
+        _vexp_y = v_fma(_vexp_y, _vexp_x, _vexp_p4_f32);
+        _vexp_y = v_fma(_vexp_y, _vexp_x, _vexp_p5_f32);
+
+        _vexp_y = v_fma(_vexp_y, _vexp_xx, _vexp_x);
+        _vexp_y = v_add(_vexp_y, _vexp_one_fp32);
+        return v_mul(_vexp_y, v_reinterpret_as_f32(_vexp_mm));
+    }
+
+    inline v_float32 v_erf(v_float32 v) {
+        v_float32 coef0 = vx_setall_f32(c_erf_coef0),
+                  coef1 = vx_setall_f32(c_erf_coef1),
+                  coef2 = vx_setall_f32(c_erf_coef2),
+                  coef3 = vx_setall_f32(c_erf_coef3),
+                  coef4 = vx_setall_f32(c_erf_coef4),
+                  coef5 = vx_setall_f32(c_erf_coef5);
+        v_float32 ones = vx_setall_f32(1.0f),
+                  neg_zeros = vx_setall_f32(-0.f),
+                  t = v_abs(v);
+        // sign(v)
+        v_float32 sign_mask = v_and(neg_zeros, v);
+
+        t = v_div(ones, v_fma(coef0, t, ones));
+        v_float32 r = v_fma(coef1, t, coef2);
+        r = v_fma(r, t, coef3);
+        r = v_fma(r, t, coef4);
+        r = v_fma(r, t, coef5);
+        // - v * v
+        v_float32 pow_2 = v_mul(v, v);
+        v_float32 neg_pow_2 = v_xor(neg_zeros, pow_2);
+        // - exp(- v * v)
+        v_float32 exp = v_exp(neg_pow_2);
+        v_float32 neg_exp = v_xor(neg_zeros, exp);
+        v_float32 res = v_mul(t, neg_exp);
+        res = v_fma(r, res, ones);
+        return v_xor(sign_mask, res);
+    }
+#endif
+}
 
-    explicit GeluFunctor() {}
+struct GeluFunctor : public BaseFunctor {
+    using Layer = GeluLayer;
+    int vlanes;
 
-    bool supportBackend(int backendId, int)
-    {
+    explicit GeluFunctor() {
+#if (CV_SIMD || CV_SIMD_SCALABLE)
+    vlanes = VTraits<v_float32>::vlanes();
+#else
+    vlanes = 1;
+#endif
+    }
+
+    bool supportBackend(int backendId, int) {
         return backendId == DNN_BACKEND_OPENCV || backendId == DNN_BACKEND_CUDA;
     }
 
-    inline float calculate(float x) const
-    {
-        return 0.5f * x * (1.0f + erf(x * M_SQRT1_2));
+    void apply(const float* srcptr, float* dstptr, int stripeStart, int len, size_t planeSize, int cn0, int cn1) const {
+        CV_UNUSED(stripeStart);
+        for (int cn = cn0; cn < cn1; cn++, srcptr += planeSize, dstptr += planeSize) {
+            int i = 0;
+#if (CV_SIMD || CV_SIMD_SCALABLE)
+            // 0.5f * x * (1.0f + erf(x * M_SQRT1_2));
+            v_float32 half = vx_setall_f32(0.5f),
+                      one = vx_setall_f32(1.0f),
+                      reciprocal_sqrt2 = vx_setall_f32(M_SQRT1_2);
+            for (; i <= len - vlanes * 4; i += vlanes * 4) {
```

Comment:
it's quite heavy function inside. 1) does it make sense to unroll the loop by vlanes*4? 2) with such aggressive unrolling there is a risk that we will have very long tail which will slow down performance significantly. I suggest to make unrolling less aggressive. 2) it's element-wise operation. Often in NCHW the product of H and W may be quite odd number far from power-of-two. You should process `(cn1-cn0)*planeSize` as a single 1D array, so that you will have just one tail, not many tiles

Reply:
I tested with unrolling by 2 but did not observe significant difference.

---

> You should process (cn1-cn0)*planeSize as a single 1D array, so that you will have just one tail, not many tiles

It is not completely correct. The loops and parallel look like the following:

```
for b in batch:
  for c in channel:
    for i in h * w: (this most inner loop is parallelled by the number of threads)
      // ...
```

So for each thread the workload should be `b * c * stripeSize` (planeSize is the step to the next segment), which is `(cn1 - cn0) * len` using terms in the code.

This parallelism is used across all activations in the file.